[AMDGPU] Introduce ordering parameter to atomic intrinsics and introduce new llvm.amdgcn.image.atomic.load intrinsic.

sstipano opened the pull request:
This intrinsic should behave mostly identically to an llvm.amdgcn.image.load, except that:
    - It is not marked as IntrReadMem. This is to ensure that the implied memory semantics are preserved.
    - When lowering, it's MachineMemOperand is to get the "acquire" memory semantics. New intrinsic atomic.load does not implicitly have acquire semantics
   
This adds ordering value to the cachepolicy parameter to all existing atomic intrinsics. MachineMemOperand now has appropriate ordering for all amdgpu image atomics.

perlfu commented:
I think the PR description needs to be clearer:
- This adds ordering parameters to all *existing* atomic intrinsics. So this will require changes in LLVM IR code generators.
- New intrinsic atomic.load does not implicitly have acquire semantics, this is just the ordering provided in the lit test.

If ordering parameter is to be added to all intrinsics then I suspect there needs to be more test coverage that the full range of these values make their way to MIR and manifest the expected effects.

sstipanovic commented:
> I think the PR description needs to be clearer:
> 
> * This adds ordering parameters to all _existing_ atomic intrinsics. So this will require changes in LLVM IR code generators.
> * New intrinsic atomic.load does not implicitly have acquire semantics, this is just the ordering provided in the lit test.
> 
> If ordering parameter is to be added to all intrinsics then I suspect there needs to be more test coverage that the full range of these values make their way to MIR and manifest the expected effects.

Initially I was about to add ordering parameter only for the new intrinsic, but I noticed that the ordering is not set for other atomics either, even though llpc sets the ordering for atomic intrinsics. Maybe, with this change, llpc won't have to add fences anymore.
If this approach is ok, I'm happy to add more tests.

arsenm reviewed (commented):
The problem with doing this is generic passes that should really know this is an atomic operation do not know it. It sort of works for codegen, but it's also a different representation from a normal scope. In the past I've thought about introducing a call atomic flag to allow generically carrying the ordering and scope

nhaehnle commented:
> The problem with doing this is generic passes that should really know this is an atomic operation do not know it. It sort of works for codegen, but it's also a different representation from a normal scope. In the past I've thought about introducing a call atomic flag to allow generically carrying the ordering and scope

Just based off of the function attributes, generic passes should assume that there may be an atomic operation in there based on the `memory` attribute (or lack thereof). Or maybe it's a lack of `nosync`[0] -- do does this intrinsic lack `nosync` as defined?  -- I've never heard an entirely convincing story about how all this is supposed to work.

If there are passes that make assumptions because this is an intrinsic, perhaps those passes should be fixed?

[0] Actually... `nosync` is a default attribute, and we define `AMDGPUImageDimIntrinsic` as a `DefaultAttrsIntrinsic`. That seems like an already existing bug for the image_atomic intrinsics. @sstipanovic can you fix that?

sstipanovic commented:
> [0] Actually... `nosync` is a default attribute, and we define `AMDGPUImageDimIntrinsic` as a `DefaultAttrsIntrinsic`. That seems like an already existing bug for the image_atomic intrinsics. @sstipanovic can you fix that?
Fixed. Atomics don't have nosync anymore.

nhaehnle commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -806,13 +810,15 @@ class AMDGPUImageDimIntrinsic<AMDGPUDimProfile P_,
       [llvm_i32_ty,                              // texfailctrl(imm; bit 0 = tfe, bit 1 = lwe)
-       llvm_i32_ty]),                            // cachepolicy(imm; bit 0 = glc, bit 1 = slc, bit 2 = dlc)
+       llvm_i32_ty],                             // cachepolicy(imm; bit 0 = glc, bit 1 = slc, bit 2 = dlc)
+      !if(!or(P_.IsAtomic, P_.IsAtomicLoad), [llvm_i32_ty], [])),       // atomic ordering
```
We should try to avoid breakage caused by changing the function signature of these intrinsics. How about adding the ordering to the cachepolicy field?

sstipanovic replied:
Yes, good idea. Thanks for the suggestion.

sstipanovic replied:
Switched to using cachepolicy field instead of ordering parameter.

nhaehnle commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -680,6 +680,7 @@ class AMDGPUDimProfile<string opmod,
   bit ZCompare = false;
   bit Gradients = false;
   string LodClampMip = "";
+  bit IsAtomicLoad = false;
```
Why is a separate `IsAtomicLoad` used here? It should be possible to avoid it.

sstipanovic replied:
It was helpful to work with additional parameter. When I put the ordering in cachepolicy then it will only need one check for adding `nosync`.

nhaehnle commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -1154,6 +1154,8 @@ class TargetLoweringBase {
 
     MachineMemOperand::Flags flags = MachineMemOperand::MONone;
+
+    AtomicOrdering ordering = AtomicOrdering::NotAtomic;
```
This piecemeal addition of fields to `IntrinsicInfo` which ends up not actually capturing the whole breadth of MMO options is problematic.

Could we simply add a `MachineMemOperand *` field in here which, if `nullptr`, supersedes all the other fields?

sstipanovic replied:
I started refactoring this to add `MachineMemOperand` field, but half way through it seemed to me that the code was more complex due to that change. It would require to construct MMO in multiple places in `getTgtMemIntrinsic` which seemed unnecessary to me. I think we now have all the necessary fields to construct MMO. What do you think?

nhaehnle commented on llvm/lib/Target/AMDGPU/MIMGInstructions.td:
```diff
@@ -1553,8 +1578,10 @@ defm IMAGE_ATOMIC_DEC           : MIMG_Atomic_Renamed <mimgopc<0x16, 0x16, 0x1c>
 defm IMAGE_ATOMIC_FCMPSWAP      : MIMG_Atomic <mimgopc<MIMG.NOP, MIMG.NOP, 0x1d, MIMG.NOP>, "image_atomic_fcmpswap", 1, 1>;
 defm IMAGE_ATOMIC_FMIN          : MIMG_Atomic <mimgopc<MIMG.NOP, MIMG.NOP, 0x1e, MIMG.NOP>, "image_atomic_fmin", 0, 1>;
 defm IMAGE_ATOMIC_FMAX          : MIMG_Atomic <mimgopc<MIMG.NOP, MIMG.NOP, 0x1f, MIMG.NOP>, "image_atomic_fmax", 0, 1>;
+defm IMAGE_ATOMIC_LOAD          : MIMG_NoSampler_NoDisasm <mimgopc<0x00, 0x00, 0x00>, "image_load", 1>;
```
This is pretty hacky. "No disasm" is usually the realm of pseudo instructions, so can you make this a pseudo instruction?

sstipanovic replied:
When making `IMAGE_ATOMIC_LOAD` a pseudo instr, it seems AsmParser is always parsing image_load as `IMAGE_ATOMIC_LOAD` which when made pseudo fails on MC layer. Is there a way to make AsmParser match `IMAGE_LOAD` first instead?

nhaehnle reviewed (commented):
Thank you for making the `nosync` adjustment. Though now, it looks like `nosync` has to be added back to most of the other image intrinsics.

sstipanovic commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -806,13 +810,15 @@ class AMDGPUImageDimIntrinsic<AMDGPUDimProfile P_,
           [ImmArg<ArgIndex<AMDGPUImageDimIntrinsicEval<P_>.TexFailCtrlArgIndex>>,
-           ImmArg<ArgIndex<AMDGPUImageDimIntrinsicEval<P_>.CachePolicyArgIndex>>]),
+           ImmArg<ArgIndex<AMDGPUImageDimIntrinsicEval<P_>.CachePolicyArgIndex>>],
+          !if(!or(P_.IsAtomic, P_.IsAtomicLoad), [ImmArg<ArgIndex<AMDGPUImageDimIntrinsicEval<P_>.AtomicOrderingIndex>>], [IntrNoSync])),
```
Just for clarification, `nosync` is not missing for atomics.

sstipanovic commented:
> Thank you for making the `nosync` adjustment. Though now, it looks like `nosync` has to be added back to most of the other image intrinsics.
`nosync` was already added back to other intrinsics. I've added an inline comment

nhaehnle commented:
>  I started refactoring this to add MachineMemOperand field, but half way through it seemed to me that the code was more complex due to that change. It would require to construct MMO in multiple places in getTgtMemIntrinsic which seemed unnecessary to me. I think we now have all the necessary fields to construct MMO. What do you think?

We don't have all the fields though. For example, MMOs have both a "success ordering" and a "fail ordering" (for cmpxchg).

As for the complexity of the change, I'd worry that it's quite difficult to change all users of `getTgtMemIntrinsics` at once. Most targets have an implementation of this method. So my assumption was that anyway the legacy way of not creating the MMO inside of `getTgtMemIntrinsic` would still be supported. Having `getTgtMemIntrinsic` do the job would be optional, only for the cases where we care about it.

If creating the MMO is too cumbersome, then perhaps that side of it could be improved?

arsenm commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -796,7 +796,8 @@ class AMDGPUImageDimIntrinsicEval<AMDGPUDimProfile P_> {
                               list<IntrinsicProperty> props,
-                              list<SDNodeProperty> sdnodeprops> : DefaultAttrsIntrinsic<
+                              list<SDNodeProperty> sdnodeprops,
+                              string name = ""> : Intrinsic<
```
Changing the attributes of the existing intrinsics should be done separately

sstipanovic replied:
I'll create a PR for that.

jayfoad commented:
I think this needs to be split into separate patches: at least one adding `MMO` to `IntrinsicInfo`, one fixing NoSync on existing atomics, one adding the new atomic.

sstipanovic commented:
@nhaehnle @arsenm ping

arsenm commented on llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:
```diff
@@ -2652,8 +2652,10 @@ bool IRTranslator::translateCall(const User &U, MachineIRBuilder &MIRBuilder) {
-        MF->getMachineMemOperand(MPI, Info.flags, MemTy, Alignment, CI.getAAMetadata()));
+    MIB.addMemOperand(MF->getMachineMemOperand(
+        MPI, Info.flags, MemTy, Alignment, CI.getAAMetadata(),
+        /*Ranges*/ nullptr, /*SSID*/ SyncScope::System, Info.ordering,
```
```suggestion
        /*Ranges=*/ nullptr, /*SSID=*/ SyncScope::System, Info.ordering,
```

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -8367,15 +8369,17 @@ SDValue SelectionDAG::getMergeValues(ArrayRef<SDValue> Ops, const SDLoc &dl) {
-  MachineMemOperand *MMO =
-      MF.getMachineMemOperand(PtrInfo, Flags, Size, Alignment, AAInfo);
+  MachineMemOperand *MMO = MF.getMachineMemOperand(
+      PtrInfo, Flags, Size, Alignment, AAInfo, /*Ranges*/ nullptr,
+      /*SSID*/ SyncScope::System, Ordering, Ordering);
```
```suggestion
      PtrInfo, Flags, Size, Alignment, AAInfo, /*Ranges=*/ nullptr,
      /*SSID=*/ SyncScope::System, Ordering, Ordering);
```

arsenm commented on llvm/test/CodeGen/AMDGPU/atomic-image-load.ll:
```diff
@@ -0,0 +1,37 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -march=amdgcn -mcpu=gfx1010 < %s | FileCheck %s -check-prefix=GFX10
```
use explicit -global-isel=0 when mixing run lines

arsenm replied:
This isn't done?

sstipanovic replied:
It wasn't. Now it is.

arsenm commented on llvm/test/CodeGen/AMDGPU/atomic-image-load.ll:
```diff
@@ -0,0 +1,37 @@
+  ret void
+}
+
+declare float @llvm.amdgcn.image.atomic.load.1d.f32.i32(i32 immarg, i32, <8 x i32>, i32 immarg, i32 immarg)
```
Test the full range of types?

arsenm replied:
Or this?

arsenm reviewed (commented):
I'm wondering if we should encode the ordering with metadata or something better than an integer

sstipanovic commented:
Sorry for the late response.
> I'm wondering if we should encode the ordering with metadata or something better than an integer

Is visibility the only reason for this suggestion? @nhaehnle what do you think?

arsenm commented:
> Sorry for the late response.
> 
> > I'm wondering if we should encode the ordering with metadata or something better than an integer
> 
> Is visibility the only reason for this suggestion? @nhaehnle what do you think?

Partially. Previously I was thinking of introducing an IR change to allow arbitrary calls to carry this kind of information. i.e. either a control bit to add the atomicrmw operands, or to encode it as an operand bundle. Really deviating from how atomicrmw represents this is unfortunate

Commit:
[AMDGPU] Introduce llvm.amdgcn.image.atomic.load intrinsic.

This intrinsic should behave mostly identically to an llvm.amdgcn.image.load, except that:
    - It is not marked as IntrReadMem. This is to ensure that the implied memory semantics are preserved.
    - When lowering, it's MachineMemOperand is to get the "acquire" memory semantics.

arsenm commented:
> > Sorry for the late response.
> > > I'm wondering if we should encode the ordering with metadata or something better than an integer
> > 
> > 
> > Is visibility the only reason for this suggestion? @nhaehnle what do you think?
> 
> Partially. Previously I was thinking of introducing an IR change to allow arbitrary calls to carry this kind of information. i.e. either a control bit to add the atomicrmw operands, or to encode it as an operand bundle. Really deviating from how atomicrmw represents this is unfortunate

Are you planning on looking into alternative ways of encoding it attached to the call, or do you need this in the intrinsic?

sstipanovic commented:
> > > Sorry for the late response.
> > > > I'm wondering if we should encode the ordering with metadata or something better than an integer
> > > 
> > > 
> > > Is visibility the only reason for this suggestion? @nhaehnle what do you think?
> > 
> > 
> > Partially. Previously I was thinking of introducing an IR change to allow arbitrary calls to carry this kind of information. i.e. either a control bit to add the atomicrmw operands, or to encode it as an operand bundle. Really deviating from how atomicrmw represents this is unfortunate
> 
> Are you planning on looking into alternative ways of encoding it attached to the call, or do you need this in the intrinsic?

Sorry, I was on paternity leave. I will look into it.
